### ISSUE TITLE
Implementing rolling forecasts

### ISSUE THREAD

**Issue by arangatang** [awslabs/gluon-ts#831]
## Description

Implementing error evaluation using rolling forecasts (also known as time series cross validation) would bring more accurate error metrics to Gluon-TS. 

There are a couple of different ways to implement this:
1. Implement this to run it in memory in `backtest.py`
2. Edit predictors such that they reuse target values in datasets
3. Rerun backtest_metrics several times with different datasets
4. Create a new dataset class

Below I present these ideas in further detail.
## Method 1 edit backtest.py to run rolling forecasts in memory with multiple datasets
It would work something like this:

1. Model is trained on training data as usual
2. If flag is set for using rolling_forecasting, generate new dataset for testing
   1. for each timeseries in the test dataset 
      1. if timeseries fullfill some quality criterion (length, frequency, etc)
         1. generate rolling forecast dataset with X different "prediction windows" 
3. Generate forecasts using the predictor on the new dataset which is X times larger than the original one
4. Some metrics should be extracted based on both individual (original) timeseries and across all timeseries

Where would this be implemented in GluonTS:
Most of the changes would be made in [backtest.py](https://github.com/awslabs/gluon-ts/blob/5ecf0ea772f4109cd937c5a316304e36d38012a8/src/gluonts/evaluation/backtest.py)

1. is already handled and is a non-issue. 
2. can be implemented [here](https://github.com/awslabs/gluon-ts/blob/5ecf0ea772f4109cd937c5a316304e36d38012a8/src/gluonts/evaluation/backtest.py#L101) by truncating the same dataset multiple times and merging the results. See below for some pseudo code:

    ```python
    # define the timeseries used for forecasts
    dataset_rolling_trunc = dataset_trunc = TransformedDataset(dataset, transformations=[transform.AdhocTransform(truncate_target)])

    # X is the number of forecasts that should make up the rolling forecast window
    for step in range(X):
        
        # generate test set
        if(step == X-1):
            dataset = dataset + dataset_rolling_trunc
        
        dataset_trunc = TransformedDataset(
            dataset_trunc, transformations=[transform.AdhocTransform(truncate_target)]
        )
        dataset_rolling_trunc = dataset_rolling_trunc + dataset_trunc # This may be too slow
    ```

3. The predictor can then predict as normally
4. The evaluator class need changes
   1. It could be useful to generate error metrics on a per timeseries basis and not only for all timeseries. Thus it would need to be able to identify which series are copies. 

### cons

* High memory overhead

## Method 2: Edit predictors such that they reuse target values in datasets if a flag is used
This should introduce less memory overhead as subsets of the same timeseries will be utilized. 

Issues with this approach is:
* the same subsets would need to be used in the evaluator and in the same order for proper evaluations to be possible. Thus similar changes would need to be done for the evaluator
* Maybe this would need to be implemented on a per model basis? If so it is much more complicated
* Maybe it is enough to edit the predict methods of the different predictors?
* This approach could introduce more bugs as it requires meddling with the predictors themselves

### cons

* Depending on how the datasets are replicated it may introduce a high memory overhead.
* Could introduce bugs into existing predictors

## Method 3: Rerun backtest_metrics and average results

1. Generate rolling forecasting datasets and store to file or keep in memory depending on datset size
2. run [backtest_metrics](https://github.com/awslabs/gluon-ts/blob/5ecf0ea772f4109cd937c5a316304e36d38012a8/src/gluonts/evaluation/backtest.py#L121) for one dataset 
3. redo 2. for all remaining datasets
4. aggregate the metrics somehow (i.e. edit the Evaluator class)

Simple way, could leverage persistent datasets and in memory ones.
All one would need to do to generate the rolling forecasts would then be to combine the outputs of the aggregate metrics.
alternatively one could suppress the evaluation step with a flag and perform it after all backtests has finished.

### cons

* Equivalent to rerunning methods, i.e. not necessarily the most elegant solution

## Method 4: Create a new type of dataset

This new dataset could overwrite the iterator of a Dataset object such that the next value targets the current item but with a truncated target value. 
I have not looked into exactly how this can be done, however this should provide less memory overhead as data is reused. 
This is potentially the fastest method as it can all be done in memory and "no/little" new memory is needed. 

### cons 

* unclear

## References

https://otexts.com/fpp2/accuracy.html
https://pdfs.semanticscholar.org/af71/3d815a7caba8dff7248ecea05a5956b2a487.pdf

**Comment by lostella** [comment]
I would exclude option 2 and 3, and if I understand correctly I think that options 1 and 4 are pretty similar. Why do you say that option 1 would have a big memory overhead?

**Comment by arangatang** [comment]
Option 1 would have a bigger overhead due to it creating a larger dataset and keeping it in memory.  Thus a lot of duplicate data will be in memory. Option 4 would instead edit the in-memory dataset thus avoiding duplicated timeseries.

an example may help. 
Say we have 10 time series each with a target length of 100 in our dataset and a prediction length of 1. From this dataset we wish to calculate the rolling forecast evaluation over 20 entries for each timeseries. 

Applying option 1 would generate a dataset of approximately 10*20=200 timeseries all of which would need to be kept in memory. 

Applying option 4 would instead convert for example a ListDataset with a RollingForecastDataset which would, when iterated upon, return subsets of the original target values without copying them. Thus the RollingForecastingDataset would still only contain 10 time-series. but these 10 would be iterated upon 200 times.

**Comment by lostella** [comment]
I see, but I think that the `TransformedDataset` in option 1 has the same effect of lazily yielding transformed entries from the original dataset, so it should be very similar in terms of memory footprint if I understand coffectly

**Comment by arangatang** [comment]
Makes sense, I guess we could implement option 1 and measure the memory utilisation. If we see that the lazy execution does not handle it properly we could try option 4 instead? 

Also, I think that the dataset generation should be reachable through some method exposed in some utils file. Thus, a user who does not want to use backtest.py for some reason could still utilize rolling forecasts.

**Comment by lostella** [comment]
Exactly: option 4 being memory efficient implies that option 1 is memory efficient, because you can use whatever you do about 4 into 1. So in a way they are really the same thing, with option 1 just offering a higher level interface, so to speak.

**Comment by jaheba** [comment]
Something we should think of is that any given time-series can be part of both the training and the test-set. While the whole time-series is included in the test-set, only a part of it should be used in the training data:

```python
# train
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], ...}

# test
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], ...}
```

Currently, we remove the last `prediction_length` values from a time-series to feed the rest into the predictor and then compare the results.

The user has to ensure that the train and test-dataset do not overlap.

However, if we now use rolling evaluation, we need to limit the windows to not use seen data.

My current idea would be to use a new field (e.g. `"rolling_start"`) to mark unseen data. Alternatively, I could think of a global value, which is used across all time-series.

**Comment by lostella** [comment]
> Something we should think of is that any given time-series can be part of both the training and the test-set.

Yes, but I think there shouldn't be assumptions about any training having even happened. Models could train "online" (see `RPredictor` or any naive baseline).

My opinion is that the evaluation should be configured with
* a set of time series
* a time range in which you want to measure how accurate the `Predictor` is
* some other configurations about the "rolling" (e.g. how much data to step over from one prediction to the next)

If the `Predictor` comes out of some training procedure, then it's up to whoever trained the model to make sure the model didn't see any evaluation data during training. In the simplest way, this can just be something that given a dataset and a timestamp `ts`, returns you the dataset only up to `ts`: if your evaluation range is strictly beyond `ts`, then you're fine.

**Comment by jaheba** [comment]
> a time range in which you want to measure how accurate the Predictor is

Do you want to set this globally or on an item to item basis?

**Comment by lostella** [comment]
> Do you want to set this globally or on an item to item basis?

Globally should be an option I think. Providing a per-item evaluation range may be a bit involved to specify: the simplest way to do that would be to say "evaluate on the last N observations from each item".

**Comment by jaheba** [comment]
> the simplest way to do that would be to say "evaluate on the last N observations from each item".

Which doesn't work well with non uniform datasets.

I somewhat like the per-item definition: It clearly indicates which data to use when just looking at it. There might also be a case where one does not want to use a global value.

**Comment by lostella** [comment]
> Which doesn't work well with non uniform datasets.

It works as long as each item has at least length N.

I think a global setting "use from t0 to t1" for evaluation is the simplest thing to do, also clearly indicates which data to use when just looking at it, and most importantly *you can look at it* even if your dataset consists of 100k time series.

**Comment by arangatang** [comment]
How about if the dataset only has missing values within this evaluation range? Do we exclude this time series from the datasets then? 
i.e.

```python
# original
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, None, None, None, 14, 15], ...}

# train
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], ...}

# test
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, None, None, None], ...}
```

Also, should we require a minimum amount of training data for each time series?
This should make cold starts less of an issue when evaluating the system if I'm not mistaken.
i.e. avoid this:

```python
# original
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], ...}

# train
{"target": [1], ...}

# test
{"target": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], ...}
```

**Comment by lostella** [comment]
> How about if the dataset only has missing values within this evaluation range? Do we exclude this time series from the datasets then? 

I think the evaluation part should take care of this: it will mask evaluation metrics wherever ground truth is missing, and could report the number of evaluated predictions for a time serie (ie where ground truth was not missing).

> Also, should we require a minimum amount of training data for each time series?

I’m not sure we should be concerned with that here: cold start forecast should be a valid evaluation scenario.

Again, I think training should not be a concern here.

**Comment by arangatang** [comment]
Closing this issue as it is solved by #844.
For an example of how to generate rolling datasets see [/examples/run_rolling_forecast_backtest.py](https://github.com/awslabs/gluon-ts/blob/master/examples/run_rolling_forecast_backtest.py).

**Comment by GabrielDeza** [comment]
Does this work with multivariate targets? I am getting a `Data must be 1-dimensional` error when working with multivariate targets. I have tried using multivariate grouper before I create the rolling dataset and after but both did not produce the intended result. Here is a snippet of code for replicating my error.

```
import pandas as pd
import numpy as np
from gluonts.dataset.common import ListDataset
from gluonts.dataset.multivariate_grouper import MultivariateGrouper
from gluonts.dataset.rolling_dataset import StepStrategy, generate_rolling_dataset
indices = pd.date_range(start = "2020-01-01", end = "2020-10-30")
#creating some random data
target = np.random.randint(0, 100, size=(3,len(indices))) + np.linspace((1,10,50),(5,20,100), len(indices)).T.astype("int")

print(target.shape)
time_series_dicts = []
for time_series in target:
    time_series_dicts.append({"target": time_series, "start": indices[0]})
dataset = ListDataset(time_series_dicts, freq=indices[0].freq)
train_grouper = MultivariateGrouper(max_target_dim=3)
train_ds = train_grouper(dataset)

dataset_rolled = generate_rolling_dataset(
        dataset=train_ds,
        start_time=pd.Timestamp(indices[0], freq=indices[0].freq),
        end_time=pd.Timestamp(indices[-1], freq = indices[0].freq),
        strategy=StepStrategy(prediction_length=5, step_size = 5))
```
